### ISSUE TITLE
Add recommendation for events with unknown end time

### ISSUE THREAD

**Issue by chrisn** [w3c/me-media-timed-events#39]
@rjksmith mentioned the need for `DataCue` events in live streams where the start time is known, but the end time not known. We haven't mentioned this in the gap analysis or recommendations sections, so I propose to add it. There would need to be a mechanism to update the end time of an existing event in this case, for when the end time becomes known. I haven't checked, but is it possible for `emsg` events to have unknown end times?

**Comment by nigelmegitt** [comment]
An extra layer of complexity here is events where the end times are not known at the begin time, but become defined later.

**Comment by rjksmith** [comment]
I think there may be two separate requirements here.

My proposal is for a `TextTrackCue` (and thereby a `DataCue`) that persists until the end of the media, which is a common use case for WebVMT, i.e. annotate the map and remain visible. In the case of live streaming, the end of media time is unknown and the idea is to allow `Infinity` to represent the end of media time - a simple extension of the [existing HTML standard](
https://www.w3.org/TR/html52/semantics-embedded-content.html#offsets-into-the-media-resource) for `media.duration`. 

This is documented in [w3c/sdw issue #1106](https://github.com/w3c/sdw/issues/1106).  Please note that no update step is required for this solution.

The two-step process you describe may also be valuable, but that requirement and its use cases should be clearly identified. It may also be applicable to WebVMT, but is beyond the scope of my current proposal.

**Comment by chrisn** [comment]
[Related discussion](https://www.w3.org/2019/01/21-me-minutes.html#item04). @rjksmith If I create a cue with an `endTime` of `Infinity`, then later create another such cue, would this need to alter the `endTime` of the first cue, to avoid overlap?

**Comment by rjksmith** [comment]
@chrisn I don't think so, as both cues finish at the end of media time (`Infinity`) and I see no dependence between them.

I don't envisage the second cue cancelling the first. Multiple cues can be displayed simultaneously and independently. For example:
   1. At time 1:00, show marker A until the end.
   1. At time 2:00, show marker B until time 4:00.
   1. At time 3:00, show marker C until the end.

Does that answer your question?

**Comment by nigelmegitt** [comment]
There are use cases for both, I think. In general I agree with @rjksmith that a new cue beginning has no necessary impact on any other cues, however I also think we may need the semantic of "update the end time of an existing cue"; doing so requires a syntax for signalling, and requires cues to be appropriately identified, which they can be (but in general they probably are not). See https://www.w3.org/TR/html51/semantics-embedded-content.html#cue - the identifier is an arbitrary string.

There is an API to allow selection of "the cue(s) with this identifier": https://www.w3.org/TR/html51/semantics-embedded-content.html#dom-texttrackcuelist-getcuebyid which could potentially be used; it only returns the first result so it would be important to specify usable ids in order to make it work.

So the picture is mixed here: it's possible to do some of what is needed, but there's no syntax, for example in WebVTT, to allow the use case to be signalled. It could only be done with javascript which would have to get notification of the updates through some other mechanism.

**Comment by mavgit** [comment]
I agree with https://github.com/w3c/me-media-timed-events/issues/39#issue-418823597 that the case of unbounded duration needs to be supported.

I also agree with https://github.com/w3c/me-media-timed-events/issues/39#issuecomment-470967005 that we need to handle the case of when an event end time goes from undefined to defined.

However, I think:
1. there are more related requirements:
- event end time goes from undefined to defined (e.g. a sports game, after the undefined game time elapses, goes from undefined to a defined remaining time for post game interviews.)
- event end time goes from defined to undefined (e.g. award show running overtime.)
- event end time going from one defined time to another defined time (e.g. regularly scheduled one hour newscast extended by 30 minutes due to breaking news or one hour program cut short by unscheduled government speech)
2. a sentence in PR #46 moves from requirements to implementation:
> It should also be possible to update the cue's end time, should the end time subsequently become known.

This is one possible implementation of the above requirements. There are other possible implementations:
- define `TextTrackCue.endTime` = `infinity` as equal to the `HTMLMediaElement.duration`. This makes some sense because the [resource selection algorithm](https://html.spec.whatwg.org/#concept-media-load-algorithm) destroys all text track cues when the `HTMLMediaElement.duration` is destroyed. This also handles all requirements without ever changing `TextTrackCue.endTime`.
- Since there is an existing [event on duration change](https://html.spec.whatwg.org/#event-media-durationchange), JavaScript could change TextTrackCue.endTime on event-media-durationchange.

I'm not arguing for or against any of the implementations, but suggesting that the existing "It should be possible..." sentence be replaced by the list of 3 requirements in (1) above, since IG Notes need to stay as requirements.

3. These requirements also need to be added to the [DataCue Explainer](https://github.com/WICG/datacue/blob/master/explainer.md), which I believe doesn't include any of these requirements.

**Comment by chrisn** [comment]
I have added a new section to Recommendations in #46 to clarify the requirement to allow cues to change from bounded to unbounded, and vice versa. I don't want to introduce new examples in this part of the document, where we describing the specific requirements for the API.

But, the examples are useful, perhaps we can write them into the Use cases section. For example, coverage of a sports game could include chapterisation, with event times that change, as well as frame accurate overlay rendering.

We'll use the explainer to consider implementation alternatives, through example code. I'll update the explainer once #46 has been merged.

**Comment by rjksmith** [comment]
@mavgit Thanks for the clear use cases [above](https://github.com/w3c/me-media-timed-events/issues/39#issuecomment-494596204), which have made me realise that these can already be supported by the API, as follows:

a. The [`TextTrackCueList`](https://html.spec.whatwg.org/multipage/media.html#texttrackcuelist) is an existing interface, so cues can be read from the [`TextTrack`](https://html.spec.whatwg.org/multipage/media.html#texttrack) object.

b. The [`TextTrackCue`](https://html.spec.whatwg.org/multipage/media.html#texttrackcue) object exposes the `endTime` attribute which [can be set](https://html.spec.whatwg.org/multipage/media.html#dom-texttrackcue-endtime).

c. Hence, cues can be updated by removing, updating and then adding them back to the list. For example:
```javascript
function updateCueEnd(textTrack, cueIndex, endTime) {
  // read cue to update
  var cue = textTrack.cues[cueIndex];

  // remove cue
  textTrack.removeCue(cue);

  // update end time
  cue.endTime = endTime;

  // add updated cue
  textTrack.addCue(cue);
}
```

[Best practices for metadata text tracks](https://html.spec.whatwg.org/multipage/media.html#best-practices-for-metadata-text-tracks) describes a recommendation for handling scoring for a sports event and considers the live streaming case, which offers another good use case for updating media timed events. At the start of the match the score is '0-0', which could change to either '1-0' or '0-1' at an unknown time in the future.

**Comment by rjksmith** [comment]
I've separated the two requirements identified in this thread, and created corresponding issues in WICG/datacue as agreed in [MTE TF call on 20 May](https://www.w3.org/2019/05/20-me-minutes.html).

   * [Add cue end time representing end of media](https://github.com/WICG/datacue/issues/8)
   * [Allow cue end time to be updated](https://github.com/WICG/datacue/issues/9)

**Comment by chrisn** [comment]
I believe this issue is now resolved, having merged #46:
* Added use cases from @mavgit 
* Added requirements for handling of cues with unbounded duration
* Added requirements for the API to notify the web app when cues change position or data payload
